File: api/src/main/java/system/AbstractSystem.java
package system;

import java.util.EnumSet;
import java.util.Stack;

import log.Console;
import system.events.GameEvent;
import system.events.GameEventListener;
import system.events.GameEventType;
import system.objects.UpdatableObject;
import utils.FrogException;

public abstract class AbstractSystem implements UpdatableObject, GameEventListener
{
	protected Stack<GameEvent> eventStack;
	protected EnumSet<GameEventType> acceptedEventsTypes;
	
	public AbstractSystem()
	{
		this.eventStack = new Stack<GameEvent>();
		this.acceptedEventsTypes = EnumSet.noneOf(GameEventType.class);
	}
	
	protected AbstractSystem(GameEventType...eventsTypes)
	{
		this();
		
		for(GameEventType type : eventsTypes)
			this.acceptedEventsTypes.add(type);
	}
	
	/**
	 * Ajoute un nouvel évènement au système
	 * @param event Le nouvel évènement
	 */
	public void pushEvent(GameEvent event)
	{
		this.eventStack.push(event);
	}
	
	/**
	 * Permet de savoir si le système prend en charge le type d'évènement spécifié
	 * @param type Le type d'évènement à tester
	 * @return Vrai si le type d'évènement spécifié est pris en charge par le système, Faux dans le cas contraire
	 */
	public boolean isAcceptedEventType(GameEventType type)
	{
		return this.acceptedEventsTypes.contains(type);
	}
	
	@Override
	public void update(float delta)
	{
		while(!this.eventStack.isEmpty())
		{
			try
			{
				this.eventReceived(this.eventStack.pop());
			}
			catch(FrogException e)
			{
				Console.log.error(e);
			}
		}
	}
}

File: godot/scripts/api/system/AbstractSystem.cs
using FrogCreator.Api.Log;
using FrogCreator.Api.System.Events;
using FrogCreator.Api.System.Objects;
using FrogCreator.Api.Utils;

namespace FrogCreator.Api.System;

public abstract class AbstractSystem : IUpdatable, IGameEventListener
{
    protected Stack<GameEvent> EventStack;
    protected HashSet<GameEventType> AcceptedEventsTypes;

    public AbstractSystem()
    {
        EventStack = new Stack<GameEvent>();
        AcceptedEventsTypes = new HashSet<GameEventType>();
    }

    protected AbstractSystem(params GameEventType[] eventsTypes) : this()
    {
        foreach (var type in eventsTypes)
            AcceptedEventsTypes.Add(type);
    }

    /// <summary>
    /// Ajoute un nouvel évènement au système
    /// </summary>
    /// <param name="gameEvent">Le nouvel évènement</param>
    public void PushEvent(GameEvent gameEvent)
    {
        EventStack.Push(gameEvent);
    }

    /// <summary>
    /// Permet de savoir si le système prend en charge le type d'évènement spécifié
    /// </summary>
    /// <param name="type">Le type d'évènement à tester</param>
    /// <returns>Vrai si le type d'évènement spécifié est pris en charge par le système, Faux dans le cas contraire</returns>
    public bool IsAcceptedEventType(GameEventType type)
    {
        return AcceptedEventsTypes.Contains(type);
    }

    public virtual void Update(float delta)
    {
        while (EventStack.Count > 0)
        {
            try
            {
                EventReceived(EventStack.Pop());
            }
            catch (FrogException e)
            {
                Console.log.error(e);
            }
        }
    }

    public abstract void EventReceived(GameEvent gameEvent);
}
